File: backend/app/schemas/account_login.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from .api import ApiModel


class _ReasonRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    reason: str = Field(min_length=1, max_length=255)

    @field_validator("reason")
    @classmethod
    def require_reason_text(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("操作原因不能为空")
        return normalized


class LoginBatchPrecheckRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    pool_id: int
    lines_text: str = Field(min_length=1)


class LoginBatchBindingDecision(BaseModel):
    model_config = ConfigDict(extra="forbid")
    line_no: int
    replace_binding: bool = False
    expected_binding_version: int | None = None


class LoginBatchCreateRequest(LoginBatchPrecheckRequest):
    binding_decisions: list[LoginBatchBindingDecision] = Field(default_factory=list)
    preview_token: str = Field(min_length=1)
    preview_fingerprint: str = Field(min_length=64, max_length=64)
    idempotency_key: str = Field(min_length=8, max_length=80)
    reason: str = Field(min_length=1, max_length=255)

    @model_validator(mode="after")
    def normalize_reason(self) -> "LoginBatchCreateRequest":
        self.reason = self.reason.strip()
        if not self.reason:
            raise ValueError("操作原因不能为空")
        return self


class LoginBatchPrecheckItemOut(BaseModel):
    line_no: int
    phone_masked: str
    route_hint: str
    account_id: int | None = None
    current_pool_id: int | None = None
    code_source_note: str
    binding_action: str = "bind"
    current_binding_note: str = ""
    current_binding_version: int = 0


class LoginBatchPrecheckOut(BaseModel):
    preview_token: str
    preview_fingerprint: str
    expires_at: datetime
    total_count: int
    create_count: int
    existing_probe_required_count: int
    migrate_count: int
    queue_position: int
    estimated_seconds: int
    worst_case_seconds: int
    credential_expires_at: datetime
    items: list[LoginBatchPrecheckItemOut]


class LoginBatchItemOut(ApiModel):
    id: int
    batch_id: int
    line_no: int
    phone_masked: str
    code_source_host: str
    code_source_uuid_hint: str
    code_source_note: str = ""
    route_hint: str
    route: str
    account_id: int | None
    initialization_policy: str = "legacy_login_only"
    authorization_status: str = "not_confirmed"
    post_initialization_id: int | None = None
    post_initialization_status: str = "not_requested"
    post_initialization_failure_type: str = ""
    status: str
    phase: str
    failure_type: str
    failure_detail: str
    warning_detail: str
    current_attempt_id: int | None
    current_attempt_state_version: int = 0
    reconcile_status: str = "none"
    reconcile_attempted: bool = False
    account_binding_version: int = 0
    execution_generation: int
    retry_count: int
    state_version: int
    next_retry_at: datetime | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime
    updated_at: datetime

    @field_validator("route", mode="before")
    @classmethod
    def canonicalize_route(cls, value: str) -> str:
        return "new_account" if value == "create" else value


class LoginBatchOut(ApiModel):
    id: int
    tenant_id: int
    pool_id: int
    created_by: str
    status: str
    state_version: int
    execution_generation: int
    resolution_version: int
    total_count: int
    success_count: int
    failed_count: int
    unresolved_count: int
    warning_count: int
    skipped_count: int
    authorized_count: int = 0
    fully_initialized_count: int = 0
    post_init_waiting_count: int = 0
    manual_required_count: int = 0
    initialization_policy: str = "legacy_login_only"
    reason: str
    trace_id: str
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime


class LoginBatchDetailOut(LoginBatchOut):
    items: list[LoginBatchItemOut] = Field(default_factory=list)


class LoginBatchRetryRequest(_ReasonRequest):
    item_ids: list[int] | None = None
    expected_state_version: int
    expected_attempt_id: int | None = None
    expected_attempt_version: int | None = None
    expected_resolution_version: int | None = None
    confirm_remote_unknown: bool = False


class LoginBatchRefreshCredentialRequest(_ReasonRequest):
    code_url: str = Field(min_length=1)
    expected_item_version: int
    expected_binding_version: int | None = None
    replace_binding: bool = False


class LoginBatchCancelRequest(_ReasonRequest):
    expected_state_version: int


class LoginBatchCapabilityOut(BaseModel):
    mode: str
    post_login_init_mode: str
    max_lines: int
    worker_concurrency: int
    post_login_init_worker_concurrency: int
    item_deadline_seconds: int
    code_wait_seconds: int
    poll_interval_seconds: int
    readiness: bool
    blockers: list[str]


class LoginBatchPostInitializationOut(ApiModel):
    id: int
    account_id: int
    generation: int
    predecessor_initialization_id: int | None
    target_pool_id: int
    policy_version: str
    status: str
    stage: str
    source_two_fa_kind: str
    two_fa_status: str
    two_fa_call_state: str
    two_fa_evidence_present: bool
    profile_status: str
    profile_batch_id: int | None
    profile_action_types: list[str]
    profile_evidence_present: bool
    abc_status: str
    abc_batch_id: str
    abc_evidence_present: bool
    abc_request_id: int | None
    abc_request_status: str
    failure_type: str
    failure_detail: str
    execution_owner: str
    version: int
    next_retry_at: datetime | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime
    updated_at: datetime


class PostLoginInitializationActionRequest(_ReasonRequest):
    expected_version: int = Field(ge=1)


class PostLoginTwoFaCandidateRequest(PostLoginInitializationActionRequest):
    candidate_password: str = Field(min_length=1, max_length=255)


class PostLoginTwoFaEmailRequest(PostLoginInitializationActionRequest):
    confirmation_code: str = Field(min_length=1, max_length=32)


class LoginBatchNotificationOut(ApiModel):
    id: int
    batch_id: int
    execution_generation: int
    resolution_version: int
    event_type: str
    channel: str
    summary: dict
    delivery_status: str
    acknowledged_at: datetime | None
    state_version: int
    created_at: datetime


class LoginBatchNotificationAckRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_version: int


class CodeSourceBindingRevealRequest(_ReasonRequest):
    expected_binding_version: int


class CodeSourceBindingRevealOut(BaseModel):
    account_id: int
    host: str
    uuid: str
    binding_version: int


class PostLoginAbcPreviewRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    deployed_release_sha: str = Field(pattern="^[0-9a-f]{40}([0-9a-f]{24})?$")


class PostLoginAbcApproveRequest(PostLoginAbcPreviewRequest):
    expected_version: int = Field(ge=1)
    expected_fingerprint: str = Field(pattern="^[0-9a-f]{64}$")
    approval_ref: str = Field(min_length=1, max_length=160)


class PostLoginAbcRequestOut(ApiModel):
    id: int
    tenant_id: int
    account_id: int
    full_initialization_id: int
    status: str
    request_version: int
    requested_by: str
    approved_by: str
    approval_ref: str
    deployed_release_sha: str
    preview_fingerprint: str
    abc_batch_id: str
    failure_type: str
    failure_detail: str
    created_at: datetime
    approved_at: datetime | None
    finished_at: datetime | None


class PostLoginAbcPreviewOut(BaseModel):
    request_id: int
    request_version: int
    account_id: int
    deployed_release_sha: str
    fingerprint: str
    classification_counts: dict[str, int] = Field(default_factory=dict)


__all__ = [
    name for name in globals()
    if name.startswith("LoginBatch")
    or name.startswith("CodeSource")
    or name.startswith("PostLogin")
]

File: frontend/src/app/types/accountLogin.ts
export type AccountBatchLoginCapability = {
  mode: 'off' | 'reconcile_only' | 'enabled';
  post_login_init_mode: 'off' | 'reconcile_only' | 'enabled';
  max_lines: number;
  item_deadline_seconds: number;
  code_wait_seconds: number;
  poll_interval_seconds: number;
  worker_concurrency: number;
  post_login_init_worker_concurrency: number;
  readiness: boolean;
  blockers: string[];
};

export type AccountBatchLoginPreviewItem = {
  line_no: number;
  phone_masked: string;
  route_hint: 'create' | 'new_account' | 'existing_probe_required';
  account_id: number | null;
  current_pool_id: number | null;
  code_source_note: string;
  binding_action: 'bind' | 'keep' | 'replace_required';
  current_binding_note: string;
  current_binding_version: number;
};

export type AccountBatchLoginPreview = {
  preview_token: string;
  preview_fingerprint: string;
  expires_at: string;
  total_count: number;
  create_count: number;
  existing_probe_required_count: number;
  migrate_count: number;
  queue_position: number;
  estimated_seconds: number;
  worst_case_seconds: number;
  credential_expires_at: string;
  items: AccountBatchLoginPreviewItem[];
};

export type AccountBatchLoginItem = {
  id: number;
  batch_id: number;
  line_no: number;
  phone_masked: string;
  code_source_note: string;
  route_hint: string;
  route: string;
  account_id: number | null;
  initialization_policy: string;
  authorization_status: string;
  post_initialization_id: number | null;
  post_initialization_status: string;
  post_initialization_failure_type: string;
  status: string;
  phase: string;
  failure_type: string;
  failure_detail: string;
  warning_detail: string;
  current_attempt_id: number | null;
  current_attempt_state_version: number;
  reconcile_status: string;
  reconcile_attempted: boolean;
  account_binding_version: number;
  execution_generation: number;
  retry_count: number;
  state_version: number;
  next_retry_at: string | null;
  started_at: string | null;
  finished_at: string | null;
};

export type AccountBatchLogin = {
  id: number;
  tenant_id: number;
  pool_id: number;
  created_by: string;
  status: string;
  state_version: number;
  execution_generation: number;
  resolution_version: number;
  total_count: number;
  success_count: number;
  failed_count: number;
  unresolved_count: number;
  warning_count: number;
  skipped_count: number;
  authorized_count: number;
  fully_initialized_count: number;
  post_init_waiting_count: number;
  manual_required_count: number;
  initialization_policy: string;
  reason: string;
  trace_id: string;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  items?: AccountBatchLoginItem[];
};

export type AccountPostLoginInitialization = {
  id: number;
  account_id: number;
  generation: number;
  predecessor_initialization_id: number | null;
  target_pool_id: number;
  policy_version: string;
  status: string;
  stage: string;
  source_two_fa_kind: string;
  two_fa_status: string;
  two_fa_call_state: string;
  two_fa_evidence_present: boolean;
  profile_status: string;
  profile_batch_id: number | null;
  profile_action_types: string[];
  profile_evidence_present: boolean;
  abc_status: string;
  abc_batch_id: string;
  abc_evidence_present: boolean;
  abc_request_id: number | null;
  abc_request_status: string;
  failure_type: string;
  failure_detail: string;
  execution_owner: string;
  version: number;
  next_retry_at: string | null;
  two_fa_next_retry_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
};

export type PostLoginAbcRequest = {
  id: number;
  tenant_id: number;
  account_id: number;
  full_initialization_id: number;
  status: string;
  request_version: number;
  requested_by: string;
  approved_by: string;
  approval_ref: string;
  deployed_release_sha: string;
  preview_fingerprint: string;
  abc_batch_id: string;
  failure_type: string;
  failure_detail: string;
  created_at: string;
  approved_at: string | null;
  finished_at: string | null;
};

export type PostLoginAbcPreview = {
  request_id: number;
  request_version: number;
  account_id: number;
  deployed_release_sha: string;
  fingerprint: string;
  classification_counts: Record<string, number>;
};

export type AccountBatchLoginNoticeItem = {
  line_no: number;
  phone_masked: string;
  reason: string;
};

export type AccountBatchLoginNotification = {
  id: number;
  batch_id: number;
  execution_generation: number;
  resolution_version: number;
  event_type: 'initial' | 'correction';
  channel: 'platform';
  summary: {
    batch_id: number;
    status: string;
    counts: Record<string, number>;
    failed: AccountBatchLoginNoticeItem[];
    unresolved: AccountBatchLoginNoticeItem[];
    warning: AccountBatchLoginNoticeItem[];
    corrections?: Array<{
      line_no: number;
      phone_masked: string;
      from_status: string;
      to_status: string;
      reason: string;
    }>;
    tg_bot_delivery?: 'dead_letter';
  };
  delivery_status: string;
  acknowledged_at: string | null;
  state_version: number;
  created_at: string;
};
